File: lua/glacier/misc/image.lua
local Widget = require("snowcap.widget")
local Color = require("glacier.misc.color")

---Miscellaneous module to work with `snowcap.widget.Image`
---@class glacier.image
local image = {}

---@class glacier.image.AlphaMask
---@field x integer Width of the mask, in pixel.
---@field y integer Height of the mask, in pixel.
---@field color? snowcap.widget.Color|string Default color to use when converting the mask to an image.
---@field mask integer[] Mask definition.
local AlphaMask = {}

---Initialize a new `glacier.image.AlphaMask`
---
---@param mask glacier.image.AlphaMask
---@return glacier.image.AlphaMask
function AlphaMask:new(mask)
    mask = mask or {
        x = 0,
        y = 0,
        mask = {},
    }

    if type(mask.color) == "string" then
        mask.color = Color.from_hex(mask.color --[[@as string]])
    end

    assert(#mask.mask == mask.x * mask.y, "Invalid mask dimension")

    setmetatable(mask, self)
    self.__index = self

    return mask
end

local function _clamp(n)
    if n < 0 then
        return 0
    elseif n > 255 then
        return 255
    else
        return n
    end
end

local function invert_mask(mask)
    local ret = {}
    for _, v in ipairs(mask) do
        table.insert(ret, 255 - _clamp(v))
    end

    return ret
end

---Create a new `AlphaMask`, from by inverting the mask.
---@return glacier.image.AlphaMask
function AlphaMask:invert()
    local inverted = {
        x = self.x,
        y = self.y,
        mask = invert_mask(self.mask),
    }

    return self:new(inverted)
end

---Set the default color for this `AlphaMask`
---@param color snowcap.widget.Color|string
function AlphaMask:set_color(color)
    if type(color) == "string" then
        color = Color.from_hex(color --[[@as string]])
    end

    self.color = color
end

---Convert this AlphaMask to a `snowcap.widget.image.Handle`
---@return snowcap.widget.image.Handle
function AlphaMask:to_image_handle(color)
    color = color or self.color

    if type(color) == "string" then
        color = Color.from_hex(color --[[@as string]])
    end

    local img_data = {}

    for _, mask in ipairs(self.mask) do
        local pixel = {}

        pixel[1] = math.floor(color.red * 255)
        pixel[2] = math.floor(color.green * 255)
        pixel[3] = math.floor(color.blue * 255)
        pixel[4] = math.floor(color.alpha * _clamp(mask))

        table.insert(img_data, string.char(table.unpack(pixel)))
    end

    return {
        rgba = {
            width = self.x,
            height = self.y,
            rgba = table.concat(img_data),
        },
    }
end

---Convert this AlphaMask to a `snowcap.widget.Image`
---
---@param color? snowcap.widget.Color|string
---@return snowcap.widget.WidgetDef
function AlphaMask:to_image(color)
    return Widget.Image({
        handle = self:to_image_handle(color),
    })
end

image.AlphaMask = AlphaMask

---Create a `glacier.image.AlphaMask`.
---
---This is a shorthand for `glacier.image.AlphaMask:new(...)`
---@param ... glacier.image.AlphaMask
---@return glacier.image.AlphaMask
function image.alpha(...)
    return AlphaMask:new(...)
end

return image --[[@as glacier.image]]

File: rust/glacier/src/misc/image.rs
//! Image handling module.

use snowcap_api::widget::{
    Color,
    image::{self},
};

/// Utility type to create snowcap [`Image`]s from a mask of alpha value.
///
/// The mask is defined from an array of u8, where 0 represents full transparency.
/// It's possible to assign a default [`Color`] to the mask, which will be used if no [`Color`] is
/// passed to the `to_*` functions.
///
/// [`From`] implementations uses the default color to create the [`Image`] or [`image::Handle`].
///
/// When converting an [`AlphaMask`] to an image, the [`Color::alpha`] component is multiplied by
/// the one in the mask, while the color component are taken as-is.
///
/// [`Image`]: image::Image
#[derive(Clone)]
pub struct AlphaMask {
    width: usize,
    height: usize,
    color: Option<Color>,
    mask: Vec<u8>,
}

impl AlphaMask {
    /// Create a new [`AlphaMask`]
    pub fn new<I, V>(width: usize, height: usize, mask: I) -> Self
    where
        I: IntoIterator<Item = V>,
        V: Into<u8>,
    {
        let mask: Vec<_> = mask.into_iter().map(Into::into).collect();

        if mask.len() != (width * height) {
            panic!(
                "Invalid mask. Expected a length of {}, got {}",
                width * height,
                mask.len()
            );
        }

        Self {
            width,
            height,
            color: None,
            mask,
        }
    }

    /// Sets default color for this [`AlphaMask`]
    pub fn color(self, color: Color) -> Self {
        Self {
            color: Some(color),
            ..self
        }
    }

    /// Create a new [`AlphaMask`] by inverting every byte.
    pub fn invert(&self) -> Self {
        let Self {
            width,
            height,
            color,
            mask,
        } = self.clone();

        let mask = mask.iter().map(|v| 255 - v).collect();

        Self {
            width,
            height,
            color,
            mask,
        }
    }

    /// Create an [`image::Handle`] from this [`AlphaMask`].
    ///
    /// If `color` is set, use this instead of the default color.
    pub fn to_image_handle(&self, color: Option<Color>) -> image::Handle {
        let color = color.or(self.color).unwrap_or(Color::rgb(0.0, 0.0, 0.0));

        let red = (color.red * 255.) as u8;
        let green = (color.green * 255.) as u8;
        let blue = (color.blue * 255.) as u8;

        let bytes = self
            .mask
            .iter()
            .flat_map(|&m| [red, green, blue, (color.alpha * (m as f32)) as u8])
            .collect();

        image::Handle::Rgba {
            width: self.width as u32,
            height: self.height as u32,
            bytes,
        }
    }

    /// Create an [`image::Image`] from this [`AlphaMask`].
    ///
    /// If `color` is set, use this instead of the default color.
    pub fn to_image(&self, color: Option<Color>) -> image::Image {
        image::Image::new(self.to_image_handle(color))
    }
}

impl From<AlphaMask> for image::Handle {
    fn from(value: AlphaMask) -> Self {
        value.to_image_handle(None)
    }
}

impl From<&AlphaMask> for image::Handle {
    fn from(value: &AlphaMask) -> Self {
        value.to_image_handle(None)
    }
}

impl From<AlphaMask> for image::Image {
    fn from(value: AlphaMask) -> Self {
        value.to_image(None)
    }
}

impl From<&AlphaMask> for image::Image {
    fn from(value: &AlphaMask) -> Self {
        value.to_image(None)
    }
}
